pasture-algorithms/src/segmentation.rs

```diff
@@ -0,0 +1,313 @@
+    (line.second - line.first)
+        .cross(&(line.first - point))
+        .norm()
+        / (line.second - line.first).norm()
+}
+
+/// ransac plane algorithm in parallel
+fn ransac_plane_par<T: PointBuffer + Sync>(
+    buffer: &T,
+    distance_threshold: f64,
+    num_of_iterations: usize,
+) -> (Plane, Vec<usize>) {
+    // iterate in parallel over num_of_iterations
+    (0..num_of_iterations)
+        .into_par_iter()
+        .map(|_x| {
+            let mut rng = rand::thread_rng();
+            let rand1 = rng.gen_range(0..buffer.len());
+            let mut rand2 = rng.gen_range(0..buffer.len());
+            while rand1 == rand2 {
+                rand2 = rng.gen_range(0..buffer.len());
+            }
+            let mut rand3 = rng.gen_range(0..buffer.len());
+            // make sure we have 3 unique random numbers to generate plane model
+            while rand2 == rand3 || rand1 == rand3 {
+                rand3 = rng.gen_range(0..buffer.len());
+            }
+            let p_a: Vector3<f64> = buffer.get_attribute(&POSITION_3D, rand1);
+            let p_b: Vector3<f64> = buffer.get_attribute(&POSITION_3D, rand2);
+            let p_c: Vector3<f64> = buffer.get_attribute(&POSITION_3D, rand3);
+
+            // compute plane from the three positions
+            let vec1 = p_b - p_a;
+            let vec2 = p_c - p_a;
+            let normal = vec1.cross(&vec2);
+            let d = -normal.dot(&p_a);
+            let mut curr_hypo = Plane {
+                a: normal.x,
+                b: normal.y,
+                c: normal.z,
+                d,
+                ranking: 0,
+            };
+
+            // find all points that belong to the plane
+            let mut current_positions = vec![];
+
+            for (index, p) in buffer
+                .iter_attribute::<Vector3<f64>>(&POSITION_3D)
+                .enumerate()
+            {
+                let distance = distance_point_plane(&p, &curr_hypo);
+                if distance < distance_threshold {
+                    //we found a point that belongs to the plane
+                    curr_hypo.ranking += 1;
+                    current_positions.push(index);
+                }
+            }
+            // return the current hypothesis and the corresponding positions
+            (curr_hypo, current_positions)
+        })
+        // get the beste hypothesis from all iterations
+        .max_by(|(x, _y), (a, _b)| x.ranking.cmp(&a.ranking))
+        .unwrap()
+}
+
+/// ransac plane algorithm in serial
+fn ransac_plane_serial<T: PointBuffer + Sync>(
+    buffer: &T,
+    distance_threshold: f64,
+    num_of_iterations: usize,
+) -> (Plane, Vec<usize>) {
+    let mut best_fit = Plane {
+        a: 0.0,
+        b: 0.0,
+        c: 0.0,
+        d: 0.0,
+        ranking: 0,
+    };
+    let mut best_positions: Vec<usize> = vec![];
+
+    //iterate num_of_iterations times
+    for _i in 0..num_of_iterations {
+        let mut rng = rand::thread_rng();
```

Comment:
From here to ~line 206, the code looks identical to the one in `ransac_par`. Any chance of extracting it into a separate function to prevent code duplication?